### ISSUE TITLE
runtime/race: memory leak in race runtime

### ISSUE THREAD

**Issue by dvyukov** [golang/go#16441]
The leak was reported internally on a sever canary that runs for days.
After a day server consumes 5.6GB, after 6 days -- 12.2GB.
The leak is exposed by the following benchmark:
```go
func BenchmarkStackLeak(b *testing.B) {
	done := make(chan bool, 1)
	for i := 0; i < b.N; i++ {
		go func() {
			growStack(rand.Intn(100))
			done <- true
		}()
		<-done
	}
}

func growStack(i int) {
	if i == 0 {
		return
	}
	growStack(i - 1)
}
```
The leak is fixed upstream in :
http://llvm.org/viewvc/llvm-project/compiler-rt/trunk/lib/tsan/rtl/tsan_rtl_thread.cc?view=diff&r1=276102&r2=276103&pathrev=276103

**Comment by dvyukov** [comment]
The change is submitted. The is not autoclosed for some reason. Closing.

**Comment by dvyukov** [comment]
Reopened in case this needs to be merged into 1.7.

**Comment by bradfitz** [comment]
@dvyukov, we use separate bugs for master fixes vs branch cherry-picks now.
